### DISCUSSION TITLE
Any way to set multiple format outputs?

### DISCUSSION THREAD

**Discussion by FyZzyss** [flakeheaven/flakeheaven#130] [Q&A]
Is it possible to set multiple conclusions, as in coverage? `--cov-report xml:coverage.xml --cov-report term`
Now I get either an artifact for gitlab or a console response

**Comment by pwoolvett** [comment]
To the best of my knowledge, flake8 only admits one formatter per run.

Having said that, you can always 
* use the `--tee` flag to tee the current formatter's output to stdout.
* write your own formatter plugin to wrap two formatters...

**Reply by oliverpool** [reply]
I have the same usecase (generate a codequality report for gitlab: the JSON is for gitlab and the text version for the log).
Having two runs seems like a bit of a waste (and hacky since the exit code of the first run must be discarded).

Would you accept a PR to rework the gitlab formatter to output simple log lines to stdout if a `output-file` has be specified?

(One could think about advanced usecase to make a "TeeFormatter" which could forward all the method calls to multiple formatters, however it a bit to complex for such a simple usecase)

**Reply by pwoolvett** [reply]
Hi @oliverpool 

why using `--tee` is not enough?

btw, I'm more inclined towards removing custom formatters from flakeheaven, and having them as separate projects

**Reply by oliverpool** [reply]
> why using --tee is not enough?

Because I want the JSON output for gitlab and the text version for the logs.

> btw, I'm more inclined towards removing custom formatters from flakeheaven, and having them as separate projects

So flakeheaven would only output a JSON and the formatters would print them as gitlab json or text with colors?
This sounds quite reasonable (maybe the [codeclimate](https://github.com/codeclimate/platform/blob/master/spec/analyzers/SPEC.md#data-types)  is a good output format?)

**Reply by pwoolvett** [reply]
flakeheaven would not interfere with flake8 handling of formatters. Just depend (initially, to avoid breaking things) on a "gitlab-formatter" which would be deprecated and eventually moved from deps to optional dep.

so at first you'd just do `pip install flakeheaven`, then after a while `pip install flakeheaven[formatters]`, and finally `pip install flakeheaven flake8-gitlab-formatter` - or something similar.

I believe atm flakeheaven is trying to do too much: pylint, toml, formatters, md+rst+ipynb.


sooooo... back to the formatter thing: the ideal would be an external package and make any and all improvements there

**Reply by pwoolvett** [reply]
ps have you used https://github.com/awelzel/flake8-gl-codeclimate?

(I personally havent even used gitlab in a while...)

**Reply by oliverpool** [reply]
> ps have you used https://github.com/awelzel/flake8-gl-codeclimate?

No, I simply call `flakeheaven lint --format=gitlab --output-file gl-code-quality-report.json`.

---

I was thinking that in the future, I would do:

```
flakeheaven lint > flakeheaven.json
flakeheaven-format-codeclimate flakeheaven.json > gl-code-quality-report.json
flakeheaven-format-terminal --with-colors --with-exit-code flakeheaven.json
```

So that the `lint` command is run only once and then I can just transform the output to my liking.

What I wonder, is what schema should the `flakeheaven.json` have and if the codeclimate spec was a good schema.
